apps/aecore/src/aec_consensus_hc.erl

```diff
@@ -833,6 +845,36 @@ is_leader_valid(Node, _Trees, TxEnv, _PrevNode) ->
             aec_conductor:throw_error(parent_chain_block_not_synced)
     end.
 
+validate_pin(TxEnv, Trees) ->
+    case aec_chain_hc:pin_info({TxEnv, Trees}) of
+        undefined -> pin_missing;
+        EncTxHash ->  
+            % TODO make this code much more robust - incorrect EncTxHash, bad value from PC, incorrect hash etc.etc
+            lager:debug("PINNING: EncHash: ~p", [EncTxHash]),
+            try 
+                #{epoch := _PinEpoch, height := PinHeight, block_hash := PinHash} = 
+                    aec_parent_connector:get_pin_by_tx_hash(EncTxHash),
+                case {ok, PinHash} =:= aec_chain_state:get_key_block_hash_at_height(PinHeight) of 
+                    true -> pin_correct;
```

Comment:
1. You have a TxHash on parent chain
2. You get the transaction from parent chain (via connector and cache, hence finite)
3. The transaction on parent chain has the right format and contains a PinHeight and PinHash.
4. you check that that hash is for that height... BUT
**also to check that that is the last height of the previous epoch!**
   for example use  `epoch_info` on current height and verify that `epoch` in there is PinEpoch+1, that  `first` is `PinHeight`-1 or get the previous epoch.

Reply:
Testing against the current epoch (PinEpoch = CurEpoch - 1).